### PR TITLE
fix(ir): preserve loop-carried Out handle identity

### LINKED ISSUE #2392: [Bug] Codegen swaps loop-carried Out tensor handles
### Component

Codegen (orchestration lowering of loop-carried `pl.Out` tensors)

### Description

When an inline helper returns two same-shaped `pl.Out` tensors from a `pl.range` loop, orchestration codegen can cross-wire the loop-carried tensor handles. The generated C++ contains an adjacent two-way assignment such as:

```cpp
a__rv_v2 = b__rv_v2;
b__rv_v2 = a__rv_v2;
```

This is a handle swap, not two independent loop-carried updates. The next loop iteration therefore submits the buffers under the wrong tensor names. Compilation succeeds, but the generated orchestration is semantically incorrect.

The repro deliberately declares the FP32 output `b` before the INT32 output `a` in the helper signature and returns `b, a`. The reversed lexical/parameter order makes the issue deterministic while keeping the tensors otherwise identical in shape. The different element types make the handle mix-up easy to identify in generated code.

### What `write_two` does

`write_two` receives a dynamic row count, an FP32 input `x`, and two output tensors. For two loop iterations, each iteration launches 16 SPMD blocks. Each block owns one row and independently writes:

- `b[row, :] = x[row, :] + 1.0` (FP32),
- `a[row, :] = round(x[row, :])` (INT32).

The helper returns both `Out` values, and the outer JIT function assigns the returned pair back to `b, a`. This forces the compiler to materialize loop-carried versions for both outputs.

### Steps to Reproduce

1. Use a PyPTO checkout at commit `e2079529db77e2a726169f18d9c6e161ef51ec1b` (or a revision containing the same frontend/codegen behavior).
2. Install PyPTO and PyTorch.
3. Save the following file as `tests/repro_out_loop_carry.py`.
4. Run:

   ```bash
   python3 tests/repro_out_loop_carry.py --platform a2a3sim
   ```

5. Inspect the generated `orchestration/repro.cpp`. The script also scans all generated orchestration C++ files and exits with status 1 when it finds the adjacent swap.

```python
"""Minimal PyPTO repro for cross-wired tensor loop carries."""

import argparse
import re
from pathlib import Path

import pypto.language as pl
import torch
from pypto.runtime import RunConfig


ROWS = 32
COLS = 512
CHUNK = 16
ROUNDS = 2
ROWS_DYN = pl.dynamic("REPRO_OUT_LOOP_ROWS")


@pl.jit.inline(auto_scope=False)
def write_two(
    n: pl.Scalar[pl.INT32],
    x: pl.Tensor[[ROWS_DYN, COLS], pl.FP32],
    b: pl.Out[pl.Tensor[[ROWS_DYN, COLS], pl.FP32]],
    a: pl.Out[pl.Tensor[[ROWS_DYN, COLS], pl.INT32]],
):
    rows = pl.tensor.dim(x, 0)
    for i in pl.range(n):
        with pl.spmd(CHUNK, name_hint="repro_out_write"):
            row = i * CHUNK + pl.tile.get_block_idx()
            if row < rows:
                src = x[row : row + 1, :]
                b[row : row + 1, :] = pl.add(
                    src,
                    pl.full([1, COLS], dtype=pl.FP32, value=1.0),
                )
                a[row : row + 1, :] = pl.cast(
                    src,
                    target_type=pl.INT32,
                    mode="rint",
                )
    return b, a


@pl.jit
def repro(
    n: pl.Scalar[pl.INT32],
    x: pl.Tensor[[ROWS_DYN, COLS], pl.FP32],
    b: pl.Out[pl.Tensor[[ROWS_DYN, COLS], pl.FP32]],
    a: pl.Out[pl.Tensor[[ROWS_DYN, COLS], pl.INT32]],
):
    x.bind_dynamic(0, ROWS_DYN)
    a.bind_dynamic(0, ROWS_DYN)
    b.bind_dynamic(0, ROWS_DYN)
    b, a = write_two(n, x, b, a)
    return b, a


ASSIGNMENT = re.compile(
    r"^\s*(?P<lhs>[A-Za-z_][A-Za-z0-9_]*__rv[A-Za-z0-9_]*)\s*=\s*"
    r"(?P<rhs>[A-Za-z_][A-Za-z0-9_]*__rv[A-Za-z0-9_]*)\s*;\s*$"
)


def find_swaps(work_dir: Path):
    swaps = []
    for source in work_dir.rglob("*.cpp"):
        if "orchestration" not in source.parts:
            continue
        lines = source.read_text(errors="replace").splitlines()
        for line_no, (first, second) in enumerate(zip(lines, lines[1:]), 1):
            lhs = ASSIGNMENT.match(first)
            rhs = ASSIGNMENT.match(second)
            if lhs and rhs and lhs["lhs"] == rhs["rhs"] and lhs["rhs"] == rhs["lhs"]:
                swaps.append((source, line_no, first.strip(), second.strip()))
    return swaps


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--platform", choices=("a2a3sim", "a2a3"), default="a2a3sim")
    args = parser.parse_args()

    x = torch.arange(ROWS * COLS, dtype=torch.float32).reshape(ROWS, COLS)
    b = torch.zeros_like(x)
    a = torch.zeros((ROWS, COLS), dtype=torch.int32)
    compiled = repro.compile(
        ROUNDS,
        x,
        b,
        a,
        config=RunConfig(platform=args.platform, save_kernels=True),
    )
    work_dir = Path(compiled.output_dir)
    print(f"compiled: {work_dir}")

    swaps = find_swaps(work_dir)
    if not swaps:
        print("no adjacent __rv swap found")
        return 0

    for source, line_no, first, second in swaps:
        print(f"cross-swap: {source}:{line_no}")
        print(f"  {first}")
        print(f"  {second}")
    return 1


if __name__ == "__main__":
    raise SystemExit(main())
```

### Expected Behavior

The two loop-carried handles must remain independent. A representative generated orchestration sequence should preserve the mapping established at entry:

```cpp
Tensor b__rv_v2 = ext_b;
Tensor a__rv_v2 = ext_a;
// submit the iteration using b__rv_v2 for b and a__rv_v2 for a
// advance each handle independently to its own successor
```

The exact SSA suffixes and temporary names may differ, but there must be no assignment that maps `a` to `b` or `b` to `a`. Each output's successor must be derived from the same output's predecessor.

### Actual Behavior

On the affected revision, compilation succeeds but generated orchestration contains:

```cpp
a__rv_v2 = b__rv_v2;
b__rv_v2 = a__rv_v2;
```

The script reports this as `cross-swap` and exits 1. The two assignments are adjacent and exchange the two tensor handles. In a real multi-iteration program, subsequent task arguments can therefore bind the FP32 and INT32 outputs to the wrong buffers.

### Git Commit ID

`e2079529db77e2a726169f18d9c6e161ef51ec1b`

### NPU Kind

`N/A (frontend/orchestration codegen repro; reproduced with a2a3sim)`

### Host Platform

Linux aarch64, Python 3.11.14, PyTorch `2.11.0+cpu`

### Additional Context

Validated in the PyPTO checkout used by the DeepSeek V4 decode work. The repro does not launch device execution; it only compiles and scans generated orchestration C++. Affected output was generated under `build_output/_jit_repro_out_loop_carry_*/orchestration/repro.cpp`.

Related open issues #2234 and #1326 discuss other loop-carry or multi-output limitations, but neither reports this cross-wired two-`Out` handle assignment.

### PR DESCRIPTION
**Pull request by Little-oil**
## Summary

- Canonicalize generated InCore output contracts where the callee and caller are created together, instead of rewriting arbitrary callers later in `NormalizeReturnOrder`.
- Trace every outlined result directly to an optional parameter index with one shared parameter map. Aliases, tensor loop carries, writeback calls, forwarded tuples, and tensor `IfStmt` merges all reuse the same tracer.
- Order return expressions, output bindings, caller projections, and their types together by the generated function parameter order filtered to `Out`/`InOut`; return types are derived once from the final expressions.
- Fall back to the original whole-contract order when any result is fresh, scalar, ambiguous, untraceable, or duplicates another returned parameter.
- Remove an empty synthetic `else` after all dead phi yields are pruned, and document/test the boundary contract in English and Chinese.

## Why

Issue #2392 has captured output parameters in `b, a` order, but `ConvertToSSA` deterministically emits the internal `IfStmt` phi values in lexical `a, b` order. The outliner previously let that internal definition order become the generated return ABI. Orchestration then projected those results back onto the loop carries positionally and produced a cross-wire such as `a = b; b = a`.

The output ABI now has one owner: `ScopeOutliner`, which creates both the outlined function and its call site. Internal SSA order may remain `a, b`; the generated boundary is synchronized once in captured `Out`/`InOut` parameter order.

## Scope

The final diff is 9 files with 534 insertions and 112 deletions. The latest simplification commit removes a net 62 lines from the previous PR head by eliminating duplicate parameter containers, per-tuple tracing code, parallel return-type bookkeeping, and repeated test assertions. The broad whole-program caller permutation and its public pass/API documentation changes remain removed.

## Tests

- Rebased onto latest `origin/main` (`49aea216`, #2578); `range-diff` confirms all four patches are unchanged
- `cmake --build build --parallel 2`
- High-risk regression selection: 7 passed
- Directly affected transform/codegen suites: 197 passed after the rebase
- Full non-runtime unit suite before the patch-identical rebase: 10,168 passed, 8 skipped, 1 xfailed
- C++ tests: 1 passed
- All pre-commit hooks passed, including clang-format, cpplint, Ruff, Pyright, Markdown, and EN/ZH parity checks
- Complete post-rebase GitHub CI and Docs matrices passed, including unit, model, distributed, onboard, system, codegen, examples, pre-commit, clang-tidy, and wheel jobs
- Runtime-only local tests cannot import the machine-local stale `_task_interface.so` because it lacks `MAILBOX_FRAME_SIZE`; this existing Simpler ABI mismatch is unrelated to the patch and is already documented in `KNOWN_ISSUES.md`

Fixes #2392

### CODE REVIEW

File: src/ir/transforms/normalize_return_order_pass.cpp [hunk 1]
```diff
@@ -320,45 +333,281 @@ FunctionPtr ReorderReturns(const FunctionPtr& func, const std::vector<int>& perm
...
+TypePtr PermuteCallLikeResultType(const TypePtr& type, const std::vector<int>& permutation, bool is_submit,
+                                  const Span& span) {
+  auto tuple_type = As<TupleType>(type);
+  INTERNAL_CHECK_SPAN(tuple_type, span)
+      << "Internal error: NormalizeReturnOrder call to a reordered multi-return function has non-tuple "
+         "result type";
+
+  const size_t tail_size = is_submit ? 1 : 0;
+  INTERNAL_CHECK_SPAN(tuple_type->types_.size() == permutation.size() + tail_size, span)
+      << "Internal error: NormalizeReturnOrder call result arity does not match return permutation";
```
[the remark this answers is not in the document]
Comment (Little-oil) [reply src/ir/transforms/normalize_return_order_pass.cpp]: Rebased onto current main. This concern is superseded by the current Submit IR contract in include/pypto/ir/expr.h: Submit results are exactly the callee returns followed by TASK_ID; runtime allocation changes the alias target and does not add a result prefix. Keeping the prefix logic would now corrupt valid Submit result positions.

File: src/ir/transforms/normalize_return_order_pass.cpp [hunk 2]
```diff
@@ -320,45 +333,281 @@ FunctionPtr ReorderReturns(const FunctionPtr& func, const std::vector<int>& perm
...
+      RecordUnsafe(global_var->name_, call_expr->span_, "the reordered callee is called from an InCore body",
+                   "Move the call to a non-InCore caller, or make the callee return Out/InOut tensors in "
+                   "parameter order");
+    } else if (!directly_bound_calls_.count(call_expr)) {
+      RecordUnsafe(
+          global_var->name_, call_expr->span_, "the call result is not directly assigned to a tuple binding",
+          "Assign the call result directly to a tuple binding, then use only TupleGetItem projections "
+          "from that binding");
+    }
+  }
```
[the remark this answers is not in the document]
Comment (Little-oil) [reply src/ir/transforms/normalize_return_order_pass.cpp]: Fixed in ae48be7e. Top-level EvalStmt Call/Submit results are treated as intentionally discarded while nested reordered calls still undergo safety checks, with regressions for both forms. Group/Spmd wrappers preserve directly forwarded reordered tuples through an inverse-permutation adapter, including compiler-generated multi-value returns such as (whole tuple, out2, out1); this covers the system-tests-direct reproducer.

File: python/pypto/pypto_core/passes.pyi [hunk 3]
```diff
@@ -352,7 +352,11 @@ def fold_no_op_reshape() -> Pass:
     """Fold no-op tile.reshape assignments into Var-to-Var assignments."""
 
 def normalize_return_order() -> Pass:
-    """Create a return order normalization pass."""
+    """Create a return-order normalization pass.
+
+    Reordered Call/Submit results must be directly bound and used only through
+    tuple-element projections in non-InCore callers.
+    """
```
[the remark this answers is not in the document]
Comment (Little-oil) [reply python/pypto/pypto_core/passes.pyi]: Fixed in ae48be7e. The normalize_return_order stub now documents parameter-writeback canonicalization and InCore tuple reordering, in addition to the caller safety constraints.

File: docs/en/dev/passes/00-pass_manager.md [hunk 4]
```diff
@@ -78,7 +78,7 @@ struct PassProperties {
 | ResolveBackendOpLayouts | SSAForm, IncoreTileOps, SplitIncoreOrch, TileOps2D | SSAForm, IncoreTileOps, SplitIncoreOrch, TileOps2D, NormalizedStmtStructure | — |
 | LowerAutoVectorSplit | SSAForm, IncoreTileOps, SplitIncoreOrch, TileOps2D, TileMemoryInferred, NormalizedStmtStructure | SSAForm, IncoreTileOps, SplitIncoreOrch, TileOps2D, TileMemoryInferred, NormalizedStmtStructure | — |
 | ExpandMixedKernel | SSAForm, IncoreTileOps, SplitIncoreOrch, TileOps2D | SSAForm, MixedKernelExpanded | — |
-| NormalizeReturnOrder | SplitIncoreOrch, IncoreTileOps | — | — |
+| NormalizeReturnOrder | SplitIncoreOrch, IncoreTileOps | ReturnParamsExplicit | — |
```
[the remark this answers is not in the document]
Comment (Little-oil) [reply docs/en/dev/passes/00-pass_manager.md:133]: Resolved by rebasing onto current main: ReturnParamsExplicit is already documented in both the English and Chinese pass-property catalogs.

File: tests/ut/codegen/test_orchestration_tensor_rw.py [hunk 5]
```diff
@@ -205,6 +254,56 @@ def orch(
...
+        cross_wires = []
+        for match in carry_update_re.finditer(code):
+            rhs_base = re.match(r"(?P<base>[ab])(?:__|$)", match["rhs"])
+            if rhs_base and rhs_base["base"] != match["lhs_base"]:
+                cross_wires.append(match.group(0).strip())
+
+        assert not cross_wires, (
+            "loop-carried output handles were updated from the other output; this routes FP32 "
+            f"and INT32 buffers under the wrong names: {cross_wires}\n{code}"
+        )
```
[the remark this answers is not in the document]
Comment (Little-oil) [reply tests/ut/codegen/test_orchestration_tensor_rw.py]: Fixed in ae48be7e. The regression matcher recognizes ext_a/ext_b, underscore-suffixed aliases such as b_alias, and all emitted tensor handle declarations (Tensor, ChipTensor, and TaskTensor), so cross-wired loop carries cannot evade the assertion.

**Comment by Little-oil** [comment]
@coderabbitai review

@codex review

**Comment by Little-oil** [comment]
@coderabbitai review

@codex review

**Review by lyfne123 — APPROVED** [review]
